### PR TITLE
8271341: Opcode() != Op_If && Opcode() != Op_RangeCheck) || outcnt() == 2 assert failure with Test7179138_1.java

### PR DESCRIPTION
**Pull request by rwestrel**
The root cause of this bug is in PhaseStringOpts. In the middle of the
chain of calls that are optimized out, there's a diamond Region/If. On
most executions this diamond is optimized out by IGVN because once
PhaseStringOpts is over, all the Region's Phis are removed. But
because one input of the If/Bool/Cmp is set to top by PhaseStringOpts
when calls are removed, it sometimes happen that top propagates to the
If before the Region is optimized out. That causes control flow below
the If to become dead while it should still be reachable.

The fix I propose is to have PhaseStringOpts removed the Region/If in
that case.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8271341](https://bugs.openjdk.java.net/browse/JDK-8271341): Opcode() != Op_If && Opcode() != Op_RangeCheck) || outcnt() == 2 assert failure with Test7179138_1.java

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to 9ddafed46b1e098e20c8da06195ed33a4f4a7e85

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4944/head:pull/4944` \
`$ git checkout pull/4944`

Update a local copy of the PR: \
`$ git checkout pull/4944` \
`$ git pull https://git.openjdk.java.net/jdk pull/4944/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4944`

View PR using the GUI difftool: \
`$ git pr show -t 4944`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4944.diff">https://git.openjdk.java.net/jdk/pull/4944.diff</a>

</details>

### CODE REVIEW

**Comment by rwestrel** [comment]
Anyone for this?

**Review by vnkozlov — APPROVED** [review]
Looks good.

File: src/hotspot/share/opto/stringopts.cpp [hunk 1]
```diff
@@ -277,6 +277,19 @@ void StringConcat::eliminate_unneeded_control() {
       C->gvn_replace_by(n, n->in(0)->in(0));
       // get rid of the other projection
       C->gvn_replace_by(n->in(0)->as_If()->proj_out(false), C->top());
+    } else if (n->is_Region()) {
+      Node* iff = n->in(1)->in(0);
+      assert(iff->is_If(), "no if for the diamond");
```
Comment (vnkozlov) [inline src/hotspot/share/opto/stringopts.cpp:282]: I assume that the only Region node listed in _control list is diamond Region.
May be add assert to check that it is diamond region. Up to you.
I would like to have separate RFE to have `is_diamond_region()` - we seems check it in few places.

**Review by TobiHartmann — APPROVED** [review]
Looks good.

Did you verify that the JavaFuzzer found test that Christian attached to the bug triggers the same issue?

File: src/hotspot/share/opto/stringopts.cpp (hunk 1 above)
Comment (rwestrel) [reply src/hotspot/share/opto/stringopts.cpp:283]: I pushed an update. Is this what you had in mind?

File: src/hotspot/share/opto/stringopts.cpp [hunk 2]
```diff
@@ -277,6 +277,19 @@ void StringConcat::eliminate_unneeded_control() {
...
+      Node* bol = iff->in(1);;
```
[the remark this answers is not in the document]
Comment (rwestrel) [reply src/hotspot/share/opto/stringopts.cpp]: fixed

**Comment by rwestrel** [comment]
> Did you verify that the JavaFuzzer found test that Christian attached to the bug triggers the same issue?

That fuzzer test doesn't fail with current master. I assume it's some other issues that was fixed in the meantime.

File: src/hotspot/share/opto/stringopts.cpp (hunk 1 above)
Comment (vnkozlov) [reply src/hotspot/share/opto/stringopts.cpp:283]: yes

**Review by vnkozlov — APPROVED** [review]
Good.

**Comment by rwestrel** [comment]
Thanks for the reviews @vnkozlov @TobiHartmann

**Pull request closed without merging**